File: 04-structural-patterns/03-composite/03-composite-qry-cond.go
package main

import (
	"fmt"
	"strings"
)

type Expression interface {
	ToSQL() string
}

type Condition struct {
	Field, Operator, Value string
}

func (c *Condition) ToSQL() string {
	return fmt.Sprintf("%s %s '%s'", c.Field, c.Operator, c.Value)
}

type AndExpression struct {
	Children []Expression
}

func (a *AndExpression) ToSQL() string {
	var parts []string
	for _, child := range a.Children {
		parts = append(parts, child.ToSQL())
	}
	return "(" + strings.Join(parts, " AND ") + ")"
}

type OrExpression struct {
	Children []Expression
}

func (o *OrExpression) ToSQL() string {
	var parts []string
	for _, child := range o.Children {
		parts = append(parts, child.ToSQL())
	}
	return "(" + strings.Join(parts, " OR ") + ")"
}

func main() {
	cond1 := &Condition{"age", ">", "30"}
	cond2 := &Condition{"city", "=", "Delhi"}
	cond3 := &Condition{"salary", ">", "50000"}

	andGroup := &AndExpression{[]Expression{cond1, cond2}}
	orGroup := &OrExpression{[]Expression{andGroup, cond3}}

	fmt.Println("WHERE", orGroup.ToSQL())
}

File: 04-structural-patterns/03-composite/03-composite-qry-cond.py
from abc import ABC, abstractmethod

class Expression(ABC):
    @abstractmethod
    def to_sql(self):
        pass

class Condition(Expression):
    def __init__(self, field, operator, value):
        self.field = field
        self.operator = operator
        self.value = value

    def to_sql(self):
        return f"{self.field} {self.operator} '{self.value}'"

class AndExpression(Expression):
    def __init__(self, *children):
        self.children = children

    def to_sql(self):
        return "(" + " AND ".join(child.to_sql() for child in self.children) + ")"

class OrExpression(Expression):
    def __init__(self, *children):
        self.children = children

    def to_sql(self):
        return "(" + " OR ".join(child.to_sql() for child in self.children) + ")"

if __name__ == "__main__":
    cond1 = Condition("age", ">", "30")
    cond2 = Condition("city", "=", "Delhi")
    cond3 = Condition("salary", ">", "50000")

    and_expr = AndExpression(cond1, cond2)
    or_expr = OrExpression(and_expr, cond3)

    print("WHERE", or_expr.to_sql())
